default/python/AI/pathfinding.py

```diff
@@ -0,0 +1,139 @@
+#      are very close the target system but can't reach it due to fuel constraints
+# TODO Consider additional optimizations:
+#    - Check if universe.shortestPath complies with fuel mechanics before running
+#      pathfinder (seemingly not worth it - we have exact heuristic in that case)
+#    - Cut off branches that can't reach target due to supply distance
+#    - For large graphs, check existance on a simplified graph (use single node to represent supply clusters)
+#    - For large graphs, check if there are any must-visit nodes (e.g. only possible resupplying system),
+#      then try to find the shortest path between those and start/target.
+def find_path_with_resupply(start, target, fleet_id, minimum_fuel_at_target=0):
+    """Find the shortest possible path between two systems that complies with FreeOrion fuel mechanics.
+
+     If the fleet can travel the shortest possible path between start and target system, then return that path.
+     Otherwise, find the shortest possible detour including refueling.
+
+
+     The core algorithm is a modified A* with the universe.shortestPathDistance as heuristic.
+     While searching for a path, keep track of the fleet's fuel. Compared to standard A*/dijkstra,
+     nodes are locked only for a certain minimum level of fuel. If a longer path yields a higher fuel
+     level at a given system, then that path considered for pathfinding and added to the queue.
+
+
+    :param start: start system id
+    :type start: int
+    :param target:  target system id
+    :type target: int
+    :param fleet_id: fleet to find the path for
+    :type fleet_id: int
+    :param minimum_fuel_at_target: optional - if specified, only accept paths that leave the
+                                   fleet with at least this much fuel left at the target system
+    :type minimum_fuel_at_target: int
+    :return: shortest possible path including resupply-deroutes in the form of system ids
+             including both start and target system
+    :rtype: tuple[int]
+    """
+    path_information = namedtuple('path_information', ['distance', 'fuel', 'path'])
+    universe = fo.getUniverse()
+    empire_id = fo.empireID()
+    fleet = universe.getFleet(fleet_id)
+    supplied_systems = set(fo.getEmpire().fleetSupplyableSystemIDs)
+
+    if start == INVALID_ID or target == INVALID_ID:
+        warn("Requested path between invalid systems.")
+        return None
+
+    # make sure the target is connected to the start system
+    shortest_possible_path_distance = universe.shortestPathDistance(start, target)
+    if shortest_possible_path_distance == -1:
+        warn("Requested path between disconnected systems, doing nothing.")
+        return None
+
+    # make sure the minimum fuel at target is realistic
+    if minimum_fuel_at_target < 0:
+        error("Requested negative fuel at target.")
+        return None
+
+    if fleet.maxFuel < minimum_fuel_at_target:
```

Comment:
It is a relatively stupid query but implies no strict logic error. For example, function ```foo``` calls this function to determine whether a scout can reach the target system and then return. A query that triggers this check and the answer to it are both perfectly fine and valid in this context and definitely not an ```error```.

Should it already be caught by the caller? Probably so. That's why it is logged as a warning.

Reply:
Not necessarily stupid, though it could be valid to require the fuel check be done prior to calling this function.

Suppose the AI is checking several fleets to see which one, if any, can make it to the target with the desired remaining fuel, do we necessarily want to force the AI to separately check the maxfuel levels before calling this method, or can we simply let this method handle that aspect?    

Seems to me to be fine, slightly simpler, to let it be handled here.